Add bubble element support for TRI and TET

dhtantoy opened the pull request:
### Summary
This PR adds support for bubble elements on both TRI and TET reference cells. It enables solving the Stokes equations using the MINI element, as discussed in [1123](https://github.com/gridap/Gridap.jl/issues/1123).

### Convergence Test
I have tested the convergence order of the implementation. The test code are provided below:
```julia
function main(N)
	u((x, y)) = VectorValue(
		sin(π*x)^2*sin(2π*y),
		-sin(π*y)^2*sin(2π*x),
	)
	p((x, y)) = cos(π*x)*cos(π*y)
	f((x, y)) = VectorValue(
		-cos(π*y)*sin(π*x)*π - 2*cos(π*x)^2*sin(2π*y)*(π^2) + 6*sin(2π*y)*(sin(π*x)^2)*(π^2),
		-cos(π*x)*sin(π*y)*π + 2*cos(π*y)^2*sin(2π*x)*(π^2) - 6*sin(2π*x)*(sin(π*y)^2)*(π^2),
	)

	m = simplexify(CartesianDiscreteModel((0, 1, 0, 1), (N, N)))
	trian = Triangulation(m)
	dx = Measure(trian, 3)

	B_test = TestFESpace(m, ReferenceFE(bubble, VectorValue{2, Float64}))
	B_trial = TrialFESpace(B_test)
	U_test = TestFESpace(m, ReferenceFE(lagrangian, VectorValue{2, Float64}, 1); conformity = :H1, dirichlet_tags = "boundary")
	U_trial = TrialFESpace(U_test, u)
	P_test = TestFESpace(m, ReferenceFE(lagrangian, Float64, 1); conformity = :H1, dirichlet_tags = "boundary")
	P_trial = TrialFESpace(P_test, p)
	X = MultiFieldFESpace([U_trial, B_trial, P_test])
	Y = MultiFieldFESpace([U_test, B_test, P_trial])

	a((u, p), (v, q)) = ∫(∇(v) ⊙ ∇(u) - (∇ ⋅ v) * p + q * (∇ ⋅ u))dx
	l((v, q)) = ∫(v ⋅ f)dx

	b_a((uc, ub, p), (vc, vb, q)) = a((uc + ub, p), (vc + vb, q))
	b_l((vc, vb, q)) = l((vc + vb, q))

	op = AffineFEOperator(b_a, b_l, X, Y)
	uch, ubh, _ = solve(LUSolver(), op)
	e = uch + ubh - u
	l2_err = sqrt(sum(∫(e ⋅ e)dx))
	h1_err = sqrt(sum(∫(∇(e) ⊙ ∇(e) + e⋅e)dx))

	return [l2_err, h1_err]
end

Ns = 2 .^ (2:5)
errors = mapreduce(main, hcat, Ns)
orders = @. log2(errors[:, 1:(end-1)] / errors[:, 2:end])

@info "partition: $(Ns)"
@info "L2 norm errors: $(errors[1, :])"
@info "H1 norm errors: $(errors[2, :])"
@info "convergence order of L2 norm: $(orders[1, :])"
@info "convergence order of H1 norm: $(orders[2, :])"
```
#### Output
```julia
[ Info: partition: [4, 8, 16, 32]

[ Info: L2 norm errors: [0.18165254729039423, 0.05686892638365796, 0.01625734745395992, 0.0053042351258857125]

[ Info: H1 norm errors: [2.3322306999742337, 1.2839798115840502, 0.6630825398048837, 0.3390325680846419]

[ Info: convergence order of L2 norm: [1.6754691232952965, 1.8065486814780825, 1.6158752530382587]

[ Info: convergence order of H1 norm: [0.8610879856454763, 0.9533621469300919, 0.9677645987245991]
```

JordiManyer commented:
Hi @dhtantoy and thank you for the PR. I have some comments on generality of the implementation: 

- First, I don't think the new monomial basis is needed at all. This should be a `linear_combination` onto a regular monomial basis. 
- I also think the new element is very specific to the MINI element, and can be easily generalised so that other people can extend. 

I think the following changes should take no time: 
What I would do is to accept 3 kwargs into the `BubbleRefFE` constructor: 

```julia
BubbleRefFE(::Type{T}, p::Polytope{D};type=:mini,coeffs=nothing,terms=nothing)
```

then inside we can have something like 

```julia
if isnothing(coeffs) && isnothing(terms)
  if type == :mini 
    coeffs = [...] # Coefficients for the linear combination
    terms = [...] # Cartesian indices for the monomial terms
  else
    throw error....
end

basis = linear_combination(coeffs, MonomialBasis(T, terms))
```

I haven't checked the syntax, but something like that should work. 

Basically the behaviour is: If we want to implement new bubbles, we can hardcode the coeffs and terms and give it a name. But we leave the door open to people just giving the coeffs and terms by hand (i.e we can then create any bubble we want). 

What do you think? Can you implement these changes?

dhtantoy commented:
Hi, @JordiManyer . Thank you for the review! 
> What do you think? Can you implement these changes?

I agree with your suggestions and will implement the changes as soon as possible.



> Another thing I'm not sure about is why the addition stuff is needed. In my mind it should work out of the box, but there might be something I am missing. Can you post the error you get if you remove it?

You're right — this does work fine for `SingleFieldFEBasis`. However, it doesn't seem to work for `MultiFieldFEBasisComponent`. Here's an example that fails when I remove the addition stuff.
```julia
# works fine
# `SingleFieldFEBasis`
du = get_trial_fe_basis(U_trial)
∇(du + du)

# error occurs 
# `MultiFieldFEBasisComponent`
duc, _ = get_trial_fe_basis(X)
∇(duc + duc)
```
 
```julia
ERROR: This function is not yet implemented
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] macro expansion
    @ ~/JuliaProjects/Gridap.jl/src/Helpers/Macros.jl:21 [inlined]
  [3] return_cache(k::Broadcasting{…}, h::Gridap.Arrays.MatrixBlock{…}, f::Gridap.Arrays.MatrixBlock{…})
    @ Gridap.Fields ~/JuliaProjects/Gridap.jl/src/Fields/FieldArrayBlocks.jl:335
  [4] evaluate(::Broadcasting{…}, ::Gridap.Arrays.MatrixBlock{…}, ::Gridap.Arrays.MatrixBlock{…})
    @ Gridap.Arrays ~/JuliaProjects/Gridap.jl/src/Arrays/Maps.jl:86
  [5] return_value(k::Broadcasting{…}, h::Gridap.Arrays.MatrixBlock{…}, f::Gridap.Arrays.MatrixBlock{…})
    @ Gridap.Fields ~/JuliaProjects/Gridap.jl/src/Fields/FieldArrayBlocks.jl:331
  [6] return_type(::Broadcasting{…}, ::Gridap.Arrays.MatrixBlock{…}, ::Gridap.Arrays.MatrixBlock{…})
    @ Gridap.Arrays ~/JuliaProjects/Gridap.jl/src/Arrays/Maps.jl:62
  [7] lazy_map(::Broadcasting{…}, ::Gridap.Arrays.LazyArray{…}, ::Gridap.Arrays.LazyArray{…})
    @ Gridap.Arrays ~/JuliaProjects/Gridap.jl/src/Arrays/LazyArrays.jl:57
  [8] get_data(f::Gridap.CellData.OperationCellField{ReferenceDomain})
    @ Gridap.CellData ~/JuliaProjects/Gridap.jl/src/CellData/CellFields.jl:545
  [9] gradient(a::Gridap.CellData.OperationCellField{ReferenceDomain})
    @ Gridap.CellData ~/JuliaProjects/Gridap.jl/src/CellData/CellFields.jl:420
```

JordiManyer commented:
I'll look into the error and let you know.

JordiManyer commented:
Hi @dhtantoy , this now looks great. The only thing I would add is a test in `test/GridapTests` where you use the MINI element for Stokes, which you probably already have somewhere. 

Concerning the addition, I understand the issue and it should be fixed at some point. I however don't have time right now, so let's just commit the hack and I'll fix it whenever I have time. Since it doesn't really affect the public API, it can be fixed later. 

Finally, don't forget to pull from master and update the NEWS.md file with a small description of the changes. I'll merge after this.

dhtantoy commented:
Hi, @JordiManyer . I've pulled from master and updated the `NEWS.md`. I've also added a test file `test/GridapTests/StokesMiniTests.jl`. Please review the updates when you have time.
